erts/emulator/beam/erl_db.c

```diff
@@ -1190,6 +1190,58 @@ BIF_RETTYPE ets_take_2(BIF_ALIST_2)
     BIF_RET(ret);
 }
 
+/*
+** take_first(Tab)
+*/
+BIF_RETTYPE ets_take_first_1(BIF_ALIST_1)
+{
+    DbTable* tb;
+    int cret;
+    Eterm first,ret;
+    CHECK_TABLES();
+
+    DB_BIF_GET_TABLE(tb, DB_WRITE, LCK_WRITE_REC, BIF_ets_take_first_1);
+
+    cret = tb->common.meth->db_first(BIF_P, tb, &first);
+
+    if (cret != DB_ERROR_NONE) {
+        db_unlock(tb, LCK_WRITE_REC);
+        BIF_ERROR(BIF_P, BADARG);
+    }
+
+    cret = tb->common.meth->db_take(BIF_P, tb, first, &ret);
```

Comment:
No decision has been made yet to accept this API addition, but I just noticed that

you don't treat `'$end_of_table'` (`first==am_EOT`). It just happens to work as long as there is no such key.

Reply:
I realize now that if `first` returns am_EOT then no keys at all exist. But it could actually be a race where someone else inserts key `'$end_of_table'` and it would make the code more clear to treat `am_EOT`.

Reply:
Please note that '$end_of_table' atom can't be used as a key (https://www.erlang.org/doc/man/ets.html).
Functions take_first/1 and take_last/1 can be used to build very efficient FIFO queues using ets.